mongodb-atlas/cmd/main.go

```diff
@@ -0,0 +1,253 @@
+const costExplorerQueryFmt = "https://cloud.mongodb.com/api/atlas/v2/orgs/%s/billing/costExplorer/usage/%s"
+
+func main() {
+	fmt.Println("Initializing Mongo plugin")
+
+	configFile, err := commonconfig.GetConfigFilePath()
+	if err != nil {
+		log.Fatalf("error opening config file: %v", err)
+	}
+
+	atlasConfig, err := atlasconfig.GetAtlasConfig(configFile)
+	if err != nil {
+		log.Fatalf("error building Atlas config: %v", err)
+	}
+	log.SetLogLevel(atlasConfig.LogLevel)
+
+	// as per https://www.mongodb.com/docs/atlas/api/atlas-admin-api-ref/,
+	// atlas admin APIs have a limit of 100 requests per minute
+	rateLimiter := rate.NewLimiter(1.1, 2)
+	atlasCostSrc := AtlasCostSource{
+		rateLimiter: rateLimiter,
+		orgID:       atlasConfig.OrgID,
+	}
+	atlasCostSrc.atlasClient = getAtlasClient(*atlasConfig)
+
+	// pluginMap is the map of plugins we can dispense.
+	var pluginMap = map[string]plugin.Plugin{
+		"CustomCostSource": &ocplugin.CustomCostPlugin{Impl: &atlasCostSrc},
+	}
+
+	plugin.Serve(&plugin.ServeConfig{
+		HandshakeConfig: handshakeConfig,
+		Plugins:         pluginMap,
+		GRPCServer:      plugin.DefaultGRPCServer,
+	})
+
+}
+
+func getAtlasClient(atlasConfig atlasconfig.AtlasConfig) HTTPClient {
+	return &http.Client{
+		Transport: &digest.Transport{
+			Username: atlasConfig.PublicKey,
+			Password: atlasConfig.PrivateKey,
+		},
+	}
+}
+
+// Implementation of CustomCostSource
+type AtlasCostSource struct {
+	orgID       string
+	rateLimiter *rate.Limiter
+	atlasClient HTTPClient
+}
+
+type HTTPClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+func (a *AtlasCostSource) GetCustomCosts(req *pb.CustomCostRequest) []*pb.CustomCostResponse {
+	results := []*pb.CustomCostResponse{}
+
+	targets, err := opencost.GetWindows(req.Start.AsTime(), req.End.AsTime(), req.Resolution.AsDuration())
+	if err != nil {
+		log.Errorf("error getting windows: %v", err)
+		errResp := pb.CustomCostResponse{
+			Errors: []string{fmt.Sprintf("error getting windows: %v", err)},
+		}
+		results = append(results, &errResp)
+		return results
+	}
+
+	for _, target := range targets {
+		if target.Start().After(time.Now().UTC()) {
+			log.Debugf("skipping future window %v", target)
+			continue
+		}
+
+		log.Debugf("fetching atlas costs for window %v", target)
+		result, err := a.getAtlasCostsForWindow(&target)
+		if err != nil {
+			log.Errorf("error getting costs for window %v: %v", target, err)
+			errResp := pb.CustomCostResponse{}
+			errResp.Errors = append(errResp.Errors, fmt.Sprintf("error getting costs for window %v: %v", target, err))
+			results = append(results, &errResp)
+		} else {
+			results = append(results, result)
+		}
+	}
+
+	return results
+}
+
+func (a *AtlasCostSource) getAtlasCostsForWindow(win *opencost.Window) (*pb.CustomCostResponse, error) {
+
+	// get the token
+	token, err := CreateCostExplorerQueryToken(a.orgID, *win.Start(), *win.End(), a.atlasClient)
```

Comment:
so, the test harness exposed an issue where we can't use anything but the start of the month. WHat you will want to do here is round the window up to months. you can I think call getMonthlyWindow on this window object, however, I see it is not exported from opencost so you might just have to copy paste the code:

```
func (w Window) getMonthlyWindow() Window {
	origStart := w.Start()
	origEnd := w.End()
	// round the start and end windows to the calendar month start and ends, respectively
	roundedStart := time.Date(origStart.Year(), origStart.Month(), 1, 0, 0, 0, 0, origStart.Location())
	roundedEnd := time.Date(origEnd.Year(), origEnd.Month()+1, 1, 0, 0, 0, 0, origEnd.Location())
	// edge case - if user has exactly specified first instant of new month, does not need rounding
	if origEnd.Day() == 1 && origEnd.Hour() == 0 && origEnd.Minute() == 0 && origEnd.Second() == 0 {
		roundedEnd = *origEnd
	}
	return NewClosedWindow(roundedStart, roundedEnd)
}
```